lib/cn/cn_tree.c

```diff
@@ -1735,6 +1777,19 @@ cn_comp_commit(struct cn_compaction_work *w)
     if (split)
         cn_split_nodeids_get(w, nodeidv);
 
+    /* TODO: For now, do not let split proceed until we have removed
+     * and/or fixed all the fanout related code.
+     */
+    if (w->cw_action == CN_ACTION_SPLIT) {
+        w->cw_err = merr(EAGAIN);
+        ev_debug(1);
+        goto done;
+    }
+
+    /* TODO: Failure past this point can result in an unterminated kvset
+     * list and/or dgen conflicts on next kvdb open.  Must fix ASAP!!!
+     */
```

Comment:
This is pretty bad.  We cannot tolerate failure past this point, yet there are many cases where an error can arise.  This prevents strategies such as trying to do something and then aborting at the end if there was a conflict (e.g., TSX and LL/SC).  That was the strategy I was using to deal with overlapping split and spill, but it cannot work if we cannot abort/retry in the cn_comp_commit() path.

https://jira.micron.com/jira/browse/NFSE-5297

Reply:
There's a point of no return in this function that was documented in a comment but has been lost.  Once the "ack" record has been recorded for all kvsets, there must be no failure conditions.  The operation has been committed in CNDB, so the steps remaining to commit the changes to the CN tree must not depend on operations that can fail.  Can you add that comment back to the code?

See  the call to cndb_record_kvset_add_ack() in cn.c for an example.

Reply:
cn_comp_update_split() has allocations that can fail. This is a new error path that I added.

I expected the cndb_record_nak() in this error path to force cNDB to throw away this transaction at the time of replay. That doesn't seem to be case for some reason. Let me talk to Gaurav on this.

Reply:
That's not good.  We need to allocate all resources needed to complete the operation before writing ACKs to CNDB.

Reply:
Gaurav and I spoke about this error path last week. At that time we incorrectly assumed that cNDB replay should be able to handle a NAK for a transaction even after ACKing all the add records.

On further discussion today, we realized that there is a bug in cNDB which doesn't handle this scenario correctly. I'm going to create a ticket and let Gaurav decide on whether it needs to be fixed.

Based on the comments from Alex and Greg, I'm going to update cn_comp_update_split() to not have any error paths and push up a PR.